# <font color='red'><b>Bootstrap assignment</b> </font>

<b>There will be some functions that start with the word "grader" ex: grader_sampples(), grader_30().. etc, you should not change those function definition.

Every Grader function has to return True.</b>

<font color='blue'> <b>Importing packages</b> </font>

In [1]:
import numpy as np # importing numpy for numerical computation
from sklearn.datasets import load_boston # here we are using sklearn's boston dataset
from sklearn.metrics import mean_squared_error # importing mean_squared_error metric

In [2]:
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

In [3]:
x.shape

(506, 13)

In [4]:
x[7]

array([1.4455e-01, 1.2500e+01, 7.8700e+00, 0.0000e+00, 5.2400e-01,
       6.1720e+00, 9.6100e+01, 5.9505e+00, 5.0000e+00, 3.1100e+02,
       1.5200e+01, 3.9690e+02, 1.9150e+01])

In [5]:
y[7]

27.1

## <font color='red'><b>Task 1</b></font>

<font color='red'> <b>Step - 1</b></font>

*  <font color='blue'><b>Creating samples</b></font><br>
    <b> Randomly create 30 samples from the whole boston data points</b>
    *  Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points
    
     For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly , consider we have selected [4, 5, 7, 8, 9, 3] now we will replicate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]
* <font color='blue'><b> Create 30 samples </b></font>
    *  Note that as a part of the Bagging when you are taking the random samples <b>make sure each of the sample will have different set of columns</b><br>
Ex: Assume we have 10 columns[1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10] for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample  [7, 9, 1, 4, 5, 6, 2] and so on...
Make sure each sample will have atleast 3 feautres/columns/attributes

* <font color='red'><b> Note - While selecting the random 60% datapoints from the whole data, make sure that the selected datapoints are all exclusive, repetition is not allowed. </b></font>

<font color='red'><b>Step - 2 </b></font>

<font color='blue'><b>Building High Variance Models on each of the sample and finding train MSE value</b></font>

*  Build a regression trees on each of 30 samples.
*  Computed the predicted values of each data point(506 data points) in your corpus.
*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$
*  Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$

<font color='red'> <b>Step - 3 </b></font>

*  <font color='blue'><b>Calculating the OOB score </b></font>

*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.
*  Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

# <font color='red'><b>Task 2</b></font>

*  <font color='blue'><b>Computing CI of OOB Score and Train MSE</b></font>
  *   Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score </li>
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values (assume like a sample) find the confidence intravels of MSE and OOB Score </li>
<li> you need to report CI of MSE and CI of OOB Score </li>
<li> Note: Refer the Central_Limit_theorem.ipynb to check how to find the confidence interaval</li>
</ol>

# <font color='red'><b>Task 3</b></font>

*  <font color='blue'><b>Given a single query point predict the price of house.</b></font>

Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
Predict the house price for this point as mentioned in the step 2 of Task 1.

## <font color='red'><b>A few key points</b></font>

* Remember that the datapoints used for calculating MSE score contain some datapoints that were initially used while training the base learners (the 60% sampling). This makes these datapoints partially seen (i.e. the datapoints used for calculating the MSE score are a mixture of seen and unseen data).
Whereas, the datapoints used for calculating OOB score have only the unseen data. This makes these datapoints completely unseen and therefore appropriate for testing the model's performance on unseen data.

* Given the information above, if your logic is correct, the calculated MSE score should be less than the OOB score.

* The MSE score must lie between 0 and 10.
* The OOB score must lie between 10 and 35.

* The difference between the left nad right confidence-interval values must not be more than 10. Make sure this is true for both MSE and OOB confidence-interval values.

# <font color='blue'> <b>Task - 1</b></font>

<font color='blue'><b>Step - 1</b></font>

*  <font color='blue'> <b>Creating samples</b></font>

<font color='Orange'><b>Algorithm</b></font>

![alt text](https://i.imgur.com/OfcFrUP.jpg/)

*  <font color='blue'><b> Write code for generating samples</b></font>

In [6]:
def generating_samples(input_data, target_data):

    '''In this function, we will write code for generating 30 samples '''
    # you can use random.choice to generate random indices without replacement
    # Please have a look at this link https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html for more details
    # Please follow above pseudo code for generating samples 
    
    p = np.array(range(len(input_data)))
    q = np.array(range(3,13))
    
    Selecting_rows = np.random.choice(p, size=303, replace=False)
    Replacing_rows = np.random.choice(np.array(range(len(Selecting_rows))), size=203, replace=False)
    Selecting_columns = np.random.choice(q, np.random.randint(3,11), replace=False)
    
    sample_data = input_data[Selecting_rows[:,None], Selecting_columns]
    target_of_sample_data = target_data[Selecting_rows]
    
    #Replicating Data
    Replicating_sample_data = sample_data[Replacing_rows]
    target_of_Replicating_sample_data = target_of_sample_data[Replacing_rows]
    
    #Concatinating Data
    final_sample_data = np.vstack((sample_data, Replicating_sample_data))
    final_target_data = np.vstack((target_of_sample_data.reshape(-1,1), target_of_Replicating_sample_data.reshape(-1,1)))

    # return sampled_input_data , sampled_target_data,selected_rows,selected_columns
    #note please return as lists
    return list(final_sample_data), list(final_target_data), list(Selecting_rows), list(Selecting_columns)

<font color='cyan'> <b> Grader function - 1 </b> </fongt>

In [7]:
def grader_samples(a,b,c,d):
    length = (len(a)==506  and len(b)==506)
    sampled = (len(a)-len(set([str(i) for i in a]))==203)
    rows_length = (len(c)==303)
    column_length= (len(d)>=3)
    assert(length and sampled and rows_length and column_length)
    return True
a,b,c,d = generating_samples(x, y)
grader_samples(a,b,c,d)

True

*  <font color='blue'> <b>Create 30 samples </b>

![alt text](https://i.imgur.com/p8eZaWL.jpg)

In [8]:
# Use generating_samples function to create 30 samples 
# store these created samples in a list
list_input_data =[]
list_output_data =[]
list_selected_row= []
list_selected_columns=[]

for i in range(0,30):
    a,b,c,d = generating_samples(x,y)
    list_input_data.append(a)
    list_output_data.append(b)
    list_selected_row.append(c)
    list_selected_columns.append(d)

<font color='cyan'> <b>Grader function - 2 </b></font>

In [9]:
def grader_30(a):
    assert(len(a)==30 and len(a[0])==506)
    return True
grader_30(list_input_data)

True

<font color='red'><b>Step - 2 </b></font>

<font color='orange'><b>Flowchart for building tree</b></font>

![alt text](https://i.imgur.com/pcXfSmp.png)

*  <font color='blue'><b> Write code for building regression trees</b></font>

In [10]:
from sklearn.tree import DecisionTreeRegressor
list_of_all_models = []

for i in range(30):
    RT = DecisionTreeRegressor(max_depth=None)
    RT.fit(list_input_data[i], list_output_data[i])
    list_of_all_models.append(RT)

<font color='orange'><b>Flowchart for calculating MSE </b></font>

![alt text](https://i.imgur.com/sPEE618.png)

After getting predicted_y for each data point, we can use sklearns mean_squared_error to calculate the MSE between predicted_y and actual_y.

*  <font color='blue'><b> Write code for calculating MSE</b></font>

In [11]:
def MSE():
    y_hat = []

    for i in range(len(x)):
        array = []
        for j in range(30):
            model = list_of_all_models[j]
            predicted_y = model.predict( x[i] [list_selected_columns[j]].reshape(1, -1) )
            array.append(predicted_y)

        y_hat.append(np.median(array))
    
    mse = mean_squared_error(y, y_hat) 
    return mse

In [12]:
mse = MSE()
print(mse)

0.04133028656126481


<font color='blue'><b>Step - 3 </b></font>

<font color='orange'><b>Flowchart for calculating OOB score</b></font>

![alt text](https://i.imgur.com/95S5Mtm.png)

Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

*  <font color='blue'><b> Write code for calculating OOB score </b></font>

In [13]:
def oob_score():   
    y_pred = []

    for i in range(len(x)):
        array = []
        for j in range(30):
            if i not in list_selected_row[j]:
                model = list_of_all_models[j]
                pred = model.predict( x[i] [list_selected_columns[j]].reshape(1, -1) )
                array.append(pred)
        
        y_pred.append(np.median(array))
        
    OOB_Score = mean_squared_error(y, y_pred)
    return OOB_Score

In [14]:
OOB_score = oob_score()
print(OOB_score)

14.516542380324985


# <font color='blue'><b>Task 2</b></font>

In [15]:
from sklearn.tree import DecisionTreeRegressor
mse_list = []
oob_list = []

# Use generating_samples function to create 30 samples 
# store these created samples in a list
for i in range(35):
    list_input_data =[]
    list_output_data =[]
    list_selected_row= []
    list_selected_columns=[]

    for j in range(0,30):
        a,b,c,d = generating_samples(x,y)
        list_input_data.append(a)
        list_output_data.append(b)
        list_selected_row.append(c)
        list_selected_columns.append(d)
        
    list_of_all_models = []
    for k in range(30):
        Model = DecisionTreeRegressor()
        Model.fit(list_input_data[k], list_output_data[k])
        list_of_all_models.append(Model)
        
    
    m_s_e = MSE()
    mse_list.append(m_s_e)
    O_O_B_Score = oob_score()
    oob_list.append(O_O_B_Score)

In [16]:
# Calculating the mean
mean_mse = np.mean(mse_list)
mean_oob_score = np.mean(oob_list)

# Calculating the standard errors of mean
mse_se = np.std(mse_list) / (len(mse_list)** 0.5)
oob_score_se = np.std(oob_list) / (len(oob_list)** 0.5)

#Calculating the Confidence Interval
CI_mse = (mean_mse - 2*mse_se, mean_mse + 2*mse_se )
print('Confidence Interval of Mean: ', np.round(CI_mse,4))

CI_oob = (mean_oob_score - 2*oob_score_se, mean_oob_score + 2*oob_score_se )
print('Confidence Interval of OOB Score: ', np.round(CI_oob,4))

Confidence Interval of Mean:  [0.0965 0.1722]
Confidence Interval of OOB Score:  [13.6257 14.7248]


# <font color='blue'><b>Task 3</b></font>

<font color='orange'><b>Flowchart for Task 3</b></font>

<b>Hint: </b> We created 30 models by using 30 samples in TASK-1. Here, we need send query point "xq"  to 30 models and perform the regression on the output generated by 30 models.

![alt text](https://i.imgur.com/Y5cNhQk.png)

*  <font color='blue'><b> Write code for TASK 3 </b></font>

In [17]:
xq= np.array([0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60])

for i in range(len(x)):
    array = []
    for j in range(30):
        model = list_of_all_models[j]
        predicted_y = model.predict( xq [list_selected_columns[j]].reshape(1, -1) )
        array.append(predicted_y)

y_hat = (np.median(array))
print('Price of the House: ',y_hat)

Price of the House:  18.65


<br><br><br>

# <font color='red'><b>Observations :</b></font>

### Task 1
- We conclude that, the model is predicting the price which is closer to the actual price because we are getting very low MSE value.
- Since the model is validated on the OOB Sample, which means data hasn’t been used while training the model, so there isn’t any leakage of data and henceforth ensures a better predictive model.
- OOB_Score ensures no leakage, there is no over-fitting of the data and hence least variance.

### Task 2:

- The Confidence Interval for MSE is [0.0965-0.1722] that basically means we are 95% confident that the MSE value lies between these range.
- The Confidence Interval for OOB_Score is [13.6257-14.7248] that basically means we are 95% confident that the OOB_Score value lies between these range.

### Task 3:

- Given the query point, our task is to find the price of the house.
- By passing this query point to our model, we predict the price of house is 18.65